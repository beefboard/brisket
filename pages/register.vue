<template>
  <div class="container">
    <form @submit.prevent="register">
      <input
        v-model="username"
        v-bind:class="{ bad: !validUser }"
        @keyup="checkUsername"
        type="text"
        placeholder="Username">
      <input
        v-model="password"
        v-bind:class="{ bad: password && password2 && password != password2 }"
        type="password"
        placeholder="Password">
      <input
        v-model="password2"
        v-bind:class="{ bad: password && password2 && password != password2 }"
        type="password"
        placeholder="Retype password">
      <input
        v-model="email"
        v-bind:class="{ bad: !validEmail }"
        @keyup="checkEmail"
        type="text"
        placeholder="Email">
      <input
        v-model="firstName"
        type="text"
        placeholder="First name">
      <input
        v-model="lastName"
        type="text"
        placeholder="Last name">
      <button :disabled="!validForm || loading">Register</button>
    </form>
  </div>
</template>

<script>
import emailValidator from 'email-validator'
export default {
  data() {
    return {
      username: null,
      password: null,
      password2: null,
      firstName: null,
      lastName: null,
      email: null,
      loading: false,
      timeout: null,
      validEmail: true,
      validUser: true
    }
  },
  validate({ store, redirect }) {
    if (store.state.auth) {
      redirect('/')
      return false
    }
    return true
  },
  methods: {
    async register() {
      this.loading = true
      try {
        await this.$axios.post('/accounts', {
          username: this.username,
          password: this.password,
          email: this.email,
          firstName: this.firstName,
          lastName: this.lastName
        })
        await this.$store.dispatch('login', {
          username: this.username,
          password: this.password
        })
        await this.$router.push('/')
      } catch (e) {
        this.loading = false
        console.log(e)
        console.log('registration error')
      }
    },
    checkEmail() {
      this.validEmail = emailValidator.validate(this.email)
    },
    checkUsername() {
      if (!this.username) {
        return
      }
      clearTimeout(this.timeout)
      this.timeout = setTimeout(async () => {
        try {
          await this.$axios.get(`/accounts/${this.username}`, {
            progress: false
          })
          this.validUser = false
        } catch (e) {
          if (e.response && e.response.status == 404) {
            this.validUser = true
          }
        }
      }, 500)
    }
  },
  computed: {
    validForm() {
      if (
        !this.username ||
        !this.password ||
        !this.email ||
        !this.firstName ||
        !this.lastName
      ) {
        return false
      }

      if (!this.validUser || !this.validEmail) {
        return false
      }

      return true
    }
  }
}
</script>

<style>
.container {
}

input {
  animation: linear 0.3s ease-in-out;
}
.bad {
  border-color: red;
}
</style>
