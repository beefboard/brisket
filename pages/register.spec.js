import { shallowMount } from '@vue/test-utils'
import register from './register'
import flushPromises from 'flush-promises'

describe('login', () => {
  let mockDispatchResponse
  const mockStore = {
    dispatch: async () => {
      if (mockDispatchResponse) {
        return await mockDispatchResponse()
      }
    }
  }

  beforeEach(() => {
    mockDispatchResponse = async () => {}
  })

  test('mounts properly', () => {
    const wrapper = shallowMount(register)
    expect(wrapper.isVueInstance()).toBeTruthy()
  })

  test('renders properly', () => {
    const wrapper = shallowMount(register)
    expect(wrapper.html()).toMatchSnapshot()
  })

  it('should contain login button', () => {
    const wrapper = shallowMount(register)
    expect(wrapper.find('button')).toBeTruthy()
  })

  it('should contain a registration form', () => {
    const wrapper = shallowMount(register)
    expect(wrapper.find('form').exists()).toBe(true)
  })

  it('should highlight invalid usernames', () => {
    const wrapper = shallowMount(register, {
      mocks: {
        $store: mockStore
      }
    })
    const form = wrapper.find('form')
    const usernameInput = form.find('[placeholder="Username"]')

    usernameInput.setValue('bad username')
    usernameInput.trigger('keyup', { key: 1 })
    expect(usernameInput.classes()).toContain('bad')

    usernameInput.setValue('goodusername')
    usernameInput.trigger('keyup', { key: 1 })
    expect(usernameInput.classes()).not.toContain('bad')
  })

  it('should highlight already existing usernames', async () => {
    mockDispatchResponse = async () => {
      return { username: 'takenusername1' }
    }

    const wrapper = shallowMount(register, {
      mocks: {
        $store: mockStore
      }
    })
    const form = wrapper.find('form')
    const usernameInput = form.find('[placeholder="Username"]')

    usernameInput.setValue('takenusername')
    usernameInput.trigger('keyup', { key: 1 })

    // Wait a second for the check to be triggered
    await new Promise(resolve => {
      setTimeout(resolve, 100)
    })
    await flushPromises()

    expect(usernameInput.classes()).toContain('bad')

    mockDispatchResponse = async () => {
      throw {
        response: {
          status: 404
        }
      }
    }

    usernameInput.setValue('usernamenottaken')
    usernameInput.trigger('keyup', { key: 1 })

    // Wait a second for the check to be triggered
    await new Promise(resolve => {
      setTimeout(resolve, 100)
    })
    await flushPromises()

    expect(usernameInput.classes()).not.toContain('bad')
  })

  it('should highlight invalid emails', async () => {
    const wrapper = shallowMount(register)
    const form = wrapper.find('form')
    const emailInput = form.find('[placeholder="Email"]')

    emailInput.setValue('bad email')
    expect(emailInput.classes()).toContain('bad')

    emailInput.setValue('bademail@someaddress')
    emailInput.trigger('keyup', { key: 1 })
    expect(emailInput.classes()).toContain('bad')

    emailInput.setValue('goodemail@someaddress.com')
    emailInput.trigger('keyup', { key: 1 })
    expect(emailInput.classes()).not.toContain('bad')
  })

  test('repeated password mistype highlighting', () => {
    const wrapper = shallowMount(register)
    const form = wrapper.find('form')

    const passwordInput = form.find('[placeholder="Password"]')
    const retypePasswordInput = form.find('[placeholder="Retype password"]')

    passwordInput.setValue('password')
    retypePasswordInput.setValue('password1')

    expect(passwordInput.classes()).toContain('bad')
    expect(retypePasswordInput.classes()).toContain('bad')

    passwordInput.setValue('password')
    retypePasswordInput.setValue('password')

    expect(passwordInput.classes()).not.toContain('bad')
    expect(retypePasswordInput.classes()).not.toContain('bad')
  })

  test('form submit button should be enable when all information correct', async () => {
    const wrapper = shallowMount(register, {
      mocks: {
        $store: mockStore
      }
    })
    const form = wrapper.find('form')

    const button = form.find('button')
    expect(button.attributes('disabled')).toBe('disabled')

    form.find('[placeholder="Username"]').setValue('username')
    form.find('[placeholder="Username"]').trigger('keyup', { key: 1 })

    // Wait a second for the check to be triggered
    await new Promise(resolve => {
      setTimeout(resolve, 100)
    })
    await flushPromises()

    form.find('[placeholder="Password"]').setValue('password')
    form.find('[placeholder="Retype password"]').setValue('password')
    form.find('[placeholder="Email"]').setValue('email@email.com')
    form.find('[placeholder="First name"]').setValue('name')
    form.find('[placeholder="Last name"]').setValue('name')

    expect(button.attributes('disabled')).not.toBe('disabled')
  })

  test('form should not allow submission when incomplete', async () => {
    const wrapper = shallowMount(register, {
      mocks: {
        $store: mockStore
      }
    })
    const form = wrapper.find('form')
    form.trigger('submit')

    expect(wrapper.vm.loading).toBe(false)
  })

  test('form submit should disable button, send data to register, login, and navigate to home', async () => {
    const wrapper = shallowMount(register, {
      mocks: {
        $store: {
          dispatch: jest.fn()
        },
        $router: {
          push: jest.fn()
        }
      }
    })
    const form = wrapper.find('form')

    const button = form.find('button')
    form.find('[placeholder="Username"]').setValue('username')
    form.find('[placeholder="Username"]').trigger('keyup', { key: 1 })

    // Wait a second for the check to be triggered
    await new Promise(resolve => {
      setTimeout(resolve, 100)
    })
    await flushPromises()

    form.find('[placeholder="Password"]').setValue('password')
    form.find('[placeholder="Retype password"]').setValue('password')
    form.find('[placeholder="Email"]').setValue('email@email.com')
    form.find('[placeholder="First name"]').setValue('name1')
    form.find('[placeholder="Last name"]').setValue('name2')

    form.trigger('submit')
    expect(button.attributes('disabled')).toBe('disabled')

    await flushPromises()

    expect(wrapper.vm.$store.dispatch).toHaveBeenCalledWith('register', {
      username: 'username',
      password: 'password',
      email: 'email@email.com',
      firstName: 'name1',
      lastName: 'name2'
    })

    expect(wrapper.vm.$store.dispatch).toHaveBeenCalledWith('login', {
      username: 'username',
      password: 'password'
    })

    expect(wrapper.vm.$router.push).toHaveBeenCalledWith('/')
  })

  test('registration submission errors should not navigate, and should display error', async () => {
    let mockError = {
      response: {
        status: 500
      }
    }
    const wrapper = shallowMount(register, {
      mocks: {
        $store: {
          dispatch: async () => {
            throw mockError
          }
        },
        $router: {
          push: jest.fn()
        }
      }
    })
    const form = wrapper.find('form')

    const button = form.find('button')
    form.find('[placeholder="Username"]').setValue('username')
    form.find('[placeholder="Username"]').trigger('keyup', { key: 1 })

    // Wait a second for the check to be triggered
    await new Promise(resolve => {
      setTimeout(resolve, 100)
    })
    await flushPromises()

    form.find('[placeholder="Password"]').setValue('password')
    form.find('[placeholder="Retype password"]').setValue('password')
    form.find('[placeholder="Email"]').setValue('email@email.com')
    form.find('[placeholder="First name"]').setValue('name1')
    form.find('[placeholder="Last name"]').setValue('name2')

    form.trigger('submit')
    expect(button.attributes('disabled')).toBe('disabled')

    await flushPromises()

    expect(wrapper.vm.$router.push).not.toHaveBeenCalled()
    expect(wrapper.find('.error').html()).toContain('Server error')

    mockError = {
      response: {
        status: 422
      }
    }
    form.trigger('submit')
    await flushPromises()
    expect(wrapper.find('.error').html()).toContain('Unknown error')

    mockError = new Error('Random error')
    form.trigger('submit')
    await flushPromises()
    expect(wrapper.find('.error').html()).toContain('Connection error')
  })
})
