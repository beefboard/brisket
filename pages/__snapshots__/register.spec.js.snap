// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`login renders properly 1`] = `
<div class="container">
  <!---->
  <form><input type="text" placeholder="Username" class=""> <input type="password" placeholder="Password" class=""> <input type="password" placeholder="Retype password" class=""> <input type="text" placeholder="Email" class=""> <input type="text" placeholder="First name"> <input type="text" placeholder="Last name"> <button disabled="disabled">Register</button></form>
</div>
`;
